proposals/gc/MVP.md

```diff
@@ -0,0 +1,314 @@
+  - `const <valtype1> <: const <valtype2>`
+    - iff `<valtype1> <: <valtype2>`
+  - `var <valtype> <: var <valtype>`
+  - Note: mutable fields are *not* subtypes of immutable ones, so `const` really means constant, not read-only
+
+
+#### Defaultability
+
+* Any numeric value type is defaultable
+
+* A reference value type is defaultable if it is not of the form `ref $t`
+
+* Locals must have a type that is defaultable.
+
+* Table definitions with non-zero minimum size must have an element type that is defaultable. (Imports are not affected.)
+
+
+### Instructions
+
+#### Equality
+
+* `ref.eq` compares two references whose types support equality
+  - `ref.eq : [eqref eqref] -> [i32]`
+
+
+#### Functions
+
+* `ref.func` creates a function reference from a function index
+  - `ref.func $f : [] -> (ref $t)`
+     - iff `$f : $t`
+
+* `call_ref` calls a function through a reference
+  - `call_ref : [t1* (optref $t)] -> [t2*]`
+     - iff `$t = [t1*] -> [t2*]`
+  - traps on `null`
+
+
+#### Structures
+
+* `struct.new <typeidx>` allocates a structure of type `$t` and initialises its fields with given values
+  - `struct.new $t : [t*] -> [(ref $t)]`
+    - iff `$t = struct (mut t)*`
+
+* `struct.new_default <typeidx>` allocates a structure of type `$t` and initialises its fields with default values
+  - `struct.new_default $t : [] -> [(ref $t)]`
+    - iff `$t = struct (mut t)*`
+    - and all `t*` are defaultable
+
+* `struct.get <typeidx> <fieldidx>` reads field `$x` from a structure
+  - `struct.get $t i : [(optref $t)] -> [t]`
+    - iff `$t = struct (mut1 t1)^i (mut ti) (mut2 t2)*`
+    - and `t = unpacked(ti)`
+  - traps on `null`
+
+* `struct.set <typeidx> <fieldidx>` writes field `$x` of a structure
+  - `struct.set $t i : [(optref $t) ti] -> []`
+    - iff `$t = struct (mut1 t1)^i (var ti) (mut2 t2)*`
+    - and `t = unpacked(ti)`
+  - traps on `null`
+
+
+#### Arrays
+
+* `array.new <typeidx>` allocates an array of type `$t` and initialises its fields with a given value
+  - `array.new $t : [t i32] -> [(ref $t)]`
+    - iff `$t = array (mut t)`
+
+* `array.new_default <typeidx>` allocates an array of type `$t` and initialises its fields with the default value
+  - `array.new_default $t : [i32] -> [(ref $t)]`
+    - iff `$t = array (mut t)`
+    - and `t` is defaultable
+
+* `array.get <typeidx>` reads an element from an array
+  - `array.get $t : [(optref $t) i32] -> [t]`
+    - iff `$t = array (mut t')`
+    - and `t = unpacked(t')`
+  - traps on `null`
+
+* `array.set <typeidx>` writes an element to an array
+  - `array.set $t : [(optref $t) i32 t] -> []`
+    - iff `$t = array (var t')`
+    - and `t = unpacked(t')`
+  - traps on `null`
+
+* `array.len <typeidx>` inquires the length of an array
+  - `array.len $t : [(optref $t)] -> [i32]`
+    - iff `$t = array (mut t)`
+  - traps on `null`
+
+
+#### Integer references
+
+Tentatively, support a type of guaranteed unboxed scalars.
```

Comment:
Could you add a "TODO: this particular i31 design choice is tentative" here?

Reply:
Done.